tfjs-backend-webgpu/src/shader_preprocessor_wgsl.ts

```diff
@@ -0,0 +1,604 @@
+        let flatIndex : u32 = getOutputFlatIndex(${type}(${dims.join(', ')}));
+        setOutputFlat(flatIndex, value);
+      }
+      fn setOutputI32(${dims.map(d => `${d} : u32`).join(', ')}, value : i32) {
+        let flatIndex : u32 = getOutputFlatIndex(${type}(${dims.join(', ')}));
+        setOutputFlatI32(flatIndex, value);
+      }
+    `;
+    }
+  }
+
+  return snippet;
+}
+
+function getInputSamplingSnippet(
+    inInfo: InputInfo, outShape: number[], isVec4: boolean,
+    isFlatDispatchLayout: boolean): string {
+  let res = getSamplerFromInInfo(inInfo, isVec4);
+
+  const inShape = inInfo.shape;
+  if (inShape.length <= outShape.length) {
+    res += getSamplerAtOutputCoords(
+        inInfo, outShape, isVec4, isFlatDispatchLayout);
+  }
+
+  return res;
+}
+
+function getSamplerFromInInfo(inInfo: InputInfo, isVec4: boolean): string {
+  const texName = inInfo.name;
+  const rank = inInfo.shape.length;
+  const type = getCoordsDataTypeWgsl(rank);
+  const funcName = 'get' + texName.charAt(0).toUpperCase() + texName.slice(1);
+  const dims = ['d0', 'd1', 'd2', 'd3'].slice(0, rank);
+  const inputs = dims.map(d => `${d} : u32`).join(', ');
+
+  if (rank < 1) {
+    if (isVec4) {
+      return `
+        fn ${funcName}() -> vec4<f32>{
+          return ${texName}.numbers[0];
+        }
+      `;
+    }
+
+    return `
+      fn ${funcName}() ->f32 {
+        return ${texName}.numbers[0];
+      }
+    `;
+  }
+
+  const shapeStr =
+      `uniforms.${texName.charAt(0).toLowerCase() + texName.slice(1)}Shape`;
+  let rankStr = `${rank}`;
+  if (rank === 0) {
+    rankStr = '1';
+  }
+
+  if (isVec4) {
+    return `
+      fn ${funcName}(${inputs}) -> vec4<f32>{
+        return ${texName}.numbers[getFlatIndex${rankStr}(${type}(${
+        dims.join(',')}),
+          ${shapeStr}) / 4u];
+      }
+      `;
+  }
+
+  return `
+    fn ${funcName}(${inputs}) -> f32 {
+      return f32(${texName}.numbers[getFlatIndex${rankStr}(${type}(${
+      dims.join(',')}),
+        ${shapeStr})]);
+    }
+   `;
+}
+
+export function getSamplerAtOutputCoords(
+    inInfo: InputInfo, outShape: number[], isVec4: boolean,
+    isFlatDispatchLayout: boolean): string {
+  const texName = inInfo.name;
+  const texFuncSnippet = texName.charAt(0).toUpperCase() + texName.slice(1);
+
+  const funcName = 'get' + texFuncSnippet + 'AtOutCoords';
+
+  const inRank = inInfo.shape.length;
+  const outRank = outShape.length;
+  const type = getCoordsDataTypeWgsl(outRank);
+
+  // If the inShape equals the outShape and the dispatch layout is flat, we can
+  // directly use |gl_GlobalInvocationID.x| as the index and don't need coords
+  // conversion between these two shapes.
+  if (util.arraysEqual(inInfo.shape, outShape) && isFlatDispatchLayout) {
+    if (isVec4) {
+      return `
+        fn ${funcName}(globalId : vec3<u32>) -> vec4<f32> {
+          return ${texName}.numbers[globalId.x];
+        }
+
+        fn ${funcName}2(coords : ${type}) -> vec4<f32> {
```

Comment:
How about renaming them to below, which are more meaningful. Similar for other places
`fn ${funcName}` -> `fn ${funcName}ByGlobalId`
`fn ${funcName}2` -> `fn ${funcName}ByCoords`

Reply:
done